File: mobile/ios/App/App/CasinoSecureTransportPlugin.swift
import Foundation
import Security
import Capacitor

/// Rejects every HTTP redirect so native authentication never changes authority implicitly.
private final class CasinoNoRedirectDelegate: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}

/// Owns native HTTPS and a this-device-only Keychain session record for issue #183.
@objc(CasinoSecureTransportPlugin)
public final class CasinoSecureTransportPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CasinoSecureTransportPlugin"
    public let jsName = "CasinoSecureTransport"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "request", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "probe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "revokeAndClear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "claimDeepLink", returnType: CAPPluginReturnPromise)
    ]
    private let service = "io.github.andreivorobiev.virtualcasino.mobile-session-v1"
    private let account = "active"
    private let maximumBodyBytes = 1_048_576
    private let lock = NSLock()
    private let sessionMutation = DispatchSemaphore(value: 1)

    @objc func configure(_ call: CAPPluginCall) {
        sessionMutation.wait()
        defer { sessionMutation.signal() }
        lock.lock()
        defer { lock.unlock() }
        do {
            let backend = try exactBackend(call.getString("backendOrigin") ?? "")
            let webView = try exactWebView(call.getString("webViewOrigin") ?? "")
            var record = try loadRecord()
            if record["backend_origin"] != nil || record["webview_origin"] != nil {
                if record["backend_origin"] as? String != backend || record["webview_origin"] as? String != webView { call.reject("MOBILE_NATIVE_ORIGIN_REBIND_REJECTED") } else { call.resolve() }
                return
            }
            record["backend_origin"] = backend
            record["webview_origin"] = webView
            record["vault_generation"] = record["vault_generation"] as? Int ?? record["generation"] as? Int ?? 0
            record.removeValue(forKey: "generation")
            try saveRecord(record)
            call.resolve()
        } catch {
            call.reject("MOBILE_NATIVE_CONFIGURATION_INVALID")
        }
    }

    @objc func request(_ call: CAPPluginCall) {
        Task {
            do {
                guard let expectedGeneration = call.getInt("generation"), expectedGeneration >= 0 else { throw VaultError.staleGeneration }
                let result = try await perform(path: call.getString("path") ?? "", method: call.getString("method") ?? "GET", headers: call.getObject("headers") ?? [:], body: call.getString("body") ?? "", expectedGeneration: expectedGeneration)
                call.resolve(result)
            } catch {
                call.reject("MOBILE_NATIVE_REQUEST_FAILED")
            }
        }
    }

    @objc func probe(_ call: CAPPluginCall) {
        Task {
            sessionMutation.wait()
            defer { sessionMutation.signal() }
            do {
                let record = try loadRecord()
                guard let token = record["token"] as? String, !token.isEmpty else {
                    call.resolve(["authenticated": false, "generation": record["vault_generation"] as? Int ?? 0, "status": 401])
                    return
                }
                let generation = record["vault_generation"] as? Int ?? 0
                let result = try await perform(path: "/api/v2/auth/mobile/session", method: "GET", headers: [:], body: "", expectedGeneration: generation, serializeSession: false)
                let status = result["status"] as? Int ?? 0
                guard [200, 401].contains(status) else { throw VaultError.invalidRecord }
                if status == 200 {
                    guard let storedSessionGeneration = record["session_generation"] as? Int, storedSessionGeneration >= 1, let body = result["body"] as? String, let bodyData = body.data(using: .utf8), let envelope = try JSONSerialization.jsonObject(with: bodyData) as? [String: Any], let payload = envelope["data"] as? [String: Any], payload["authenticated"] as? Bool == true, let session = payload["session"] as? [String: Any], let sessionGeneration = session["generation"] as? Int, sessionGeneration == storedSessionGeneration, let issuedAt = session["issued_at"] as? String, !issuedAt.isEmpty, let expiresAt = session["expires_at"] as? String, !expiresAt.isEmpty, session["status"] as? String == "active" else { throw VaultError.invalidEnvelope }
                }
                call.resolve(["authenticated": status == 200, "generation": result["generation"] as? Int ?? generation, "status": status])
            } catch {
                call.reject("MOBILE_NATIVE_SESSION_PROBE_FAILED")
            }
        }
    }

    @objc func revokeAndClear(_ call: CAPPluginCall) {
        Task {
            sessionMutation.wait()
            defer { sessionMutation.signal() }
            do {
                var record = try loadRecord()
                if let token = record["token"] as? String, !token.isEmpty {
                    let generation = record["vault_generation"] as? Int ?? 0
                    let result = try await perform(path: "/api/v2/auth/mobile/session/revoke", method: "POST", headers: [:], body: "{}", expectedGeneration: generation, serializeSession: false)
                    guard [200, 401].contains(result["status"] as? Int ?? 0) else { throw VaultError.invalidRecord }
                    record = try loadRecord()
                }
                if record["token"] != nil { try clearCredentials(&record) }
                call.resolve(["revoked": true, "cleared": true])
            } catch {
                call.reject("MOBILE_NATIVE_ACCOUNT_SWITCH_REVOKE_FAILED")
            }
        }
    }

    @objc func claimDeepLink(_ call: CAPPluginCall) {
        do {
            let fingerprint = call.getString("fingerprint") ?? ""
            guard fingerprint.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil else { throw VaultError.invalidRecord }
            lock.lock()
            defer { lock.unlock() }
            var record = try loadRecord()
            var fingerprints = record["deep_link_fingerprints"] as? [String] ?? []
            guard !fingerprints.contains(fingerprint) else {
                call.resolve(["claimed": false])
                return
            }
            fingerprints = Array(fingerprints.suffix(63)) + [fingerprint]
            record["deep_link_fingerprints"] = fingerprints
            try saveRecord(record)
            call.resolve(["claimed": true])
        } catch {
            call.reject("MOBILE_NATIVE_DEEP_LINK_CLAIM_FAILED")
        }
    }

    private func perform(path: String, method: String, headers: JSObject, body: String, expectedGeneration: Int, serializeSession: Bool = true) async throws -> JSObject {
        if serializeSession { sessionMutation.wait() }
        defer { if serializeSession { sessionMutation.signal() } }
        lock.lock()
        let loadedRecord: [String: Any]
        do { loadedRecord = try loadRecord(); lock.unlock() } catch { lock.unlock(); throw error }
        var record = loadedRecord
        let vaultGeneration = record["vault_generation"] as? Int ?? 0
        guard expectedGeneration >= 0, expectedGeneration == vaultGeneration else { throw VaultError.staleGeneration }
        guard path.range(of: "^/api/(?:v1|v2)/[A-Za-z0-9_./?=&%+-]*$", options: .regularExpression) != nil, !path.contains("//") else { throw VaultError.invalidPath }
        let normalizedMethod = method.uppercased()
        guard ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"].contains(normalizedMethod) else { throw VaultError.invalidMethod }
        guard !["/api/v2/auth/login", "/api/v2/auth/guest"].contains(path) || (record["token"] as? String ?? "").isEmpty else { throw VaultError.staleGeneration }
        guard let backend = record["backend_origin"] as? String, let url = URL(string: backend + path), let webViewOrigin = record["webview_origin"] as? String else { throw VaultError.invalidRecord }
        var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData, timeoutInterval: 15)
        request.httpMethod = normalizedMethod
        request.setValue(webViewOrigin, forHTTPHeaderField: "Origin")
        request.setValue("1", forHTTPHeaderField: "X-Casino-Mobile-Client")
        let publicHeaders = Set(["accept", "content-type", "idempotency-key"])
        for (name, value) in headers {
            let lower = name.lowercased()
            guard publicHeaders.contains(lower), let stringValue = value as? String else { throw VaultError.credentialHeader }
            request.setValue(stringValue, forHTTPHeaderField: name)
        }
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        if let token = record["token"] as? String, !token.isEmpty { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if ["POST", "PUT", "PATCH", "DELETE"].contains(normalizedMethod), let csrf = record["csrf_token"] as? String, !csrf.isEmpty { request.setValue(csrf, forHTTPHeaderField: "X-CSRF-Token") }
        if let guestNonce = record["guest_browser_nonce"] as? String, !guestNonce.isEmpty { request.setValue(guestNonce, forHTTPHeaderField: "X-Guest-Browser-Nonce") }
        var requestBody = body
        if path == "/api/v2/auth/mobile/session/rotate" { requestBody = String(data: try JSONSerialization.data(withJSONObject: ["expected_generation": record["session_generation"] as? Int ?? -1]), encoding: .utf8) ?? "" }
        if !requestBody.isEmpty && ["POST", "PUT", "PATCH", "DELETE"].contains(normalizedMethod) {
            guard let bytes = requestBody.data(using: .utf8), bytes.count <= maximumBodyBytes else { throw VaultError.bodyTooLarge }
            request.httpBody = bytes
        }
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = nil
        configuration.httpShouldSetCookies = false
        configuration.urlCache = nil
        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        let session = URLSession(configuration: configuration, delegate: CasinoNoRedirectDelegate(), delegateQueue: nil)
        defer { session.finishTasksAndInvalidate() }
        let (data, response) = try await session.data(for: request)
        guard data.count <= maximumBodyBytes, let http = response as? HTTPURLResponse else { throw VaultError.bodyTooLarge }
        guard !data.isEmpty, let envelopeObject = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw VaultError.invalidEnvelope }
        var envelope = envelopeObject
        let successStatus = (200..<300).contains(http.statusCode)
        guard !successStatus || (envelopeObject["ok"] as? Bool == true && envelopeObject["data"] is [String: Any]) else { throw VaultError.invalidEnvelope }
        let terminalSession = ["/api/v2/auth/logout", "/api/v2/auth/guest/end", "/api/v2/auth/mobile/session/revoke"].contains(path)
        if terminalSession && successStatus {
            let acknowledgement = path == "/api/v2/auth/logout" ? "logged_out" : (path == "/api/v2/auth/guest/end" ? "ended" : "revoked")
            guard (envelopeObject["data"] as? [String: Any])?[acknowledgement] as? Bool == true else { throw VaultError.invalidEnvelope }
        }
        let credentialIssuance = ["/api/v2/auth/login", "/api/v2/auth/guest", "/api/v2/auth/mobile/session/rotate"].contains(path) && successStatus
        var issuedSession = false
        if credentialIssuance, var payload = envelope["data"] as? [String: Any], var sessionBody = payload["session"] as? [String: Any] {
            let token = sessionBody.removeValue(forKey: "token") as? String
            let csrf = sessionBody.removeValue(forKey: "csrf_token") as? String
            guard let token, !token.isEmpty, let csrf, !csrf.isEmpty, let sessionId = sessionBody["session_id"] as? String, !sessionId.isEmpty, let sessionGeneration = sessionBody["generation"] as? Int, sessionGeneration >= 1 else { throw VaultError.invalidEnvelope }
            record["token"] = token
            record["csrf_token"] = csrf
            record["session_id"] = sessionId
            record["session_generation"] = sessionGeneration
            record["vault_generation"] = vaultGeneration + 1
            let userBody = payload["user"] as? [String: Any]
            record["account_id"] = userBody?["user_id"] as? String ?? payload["user_id"] as? String ?? record["account_id"]
            issuedSession = true
            payload["session"] = sessionBody
            envelope["data"] = payload
        }
        guard !credentialIssuance || issuedSession else { throw VaultError.invalidEnvelope }
        var issuedGuestNonce = false
        if credentialIssuance && path == "/api/v2/auth/guest", var payload = envelope["data"] as? [String: Any] {
            let guestNonce = payload.removeValue(forKey: "guest_browser_nonce") as? String
            if let guestNonce, !guestNonce.isEmpty { record["guest_browser_nonce"] = guestNonce; issuedGuestNonce = true }
            envelope["data"] = payload
        }
        guard path != "/api/v2/auth/guest" || !credentialIssuance || issuedGuestNonce else { throw VaultError.invalidEnvelope }
        _ = try rejectCredentialResidue(envelope, depth: 0, visited: 0)
        lock.lock()
        defer { lock.unlock() }
        let current = try loadRecord()
        guard (current["vault_generation"] as? Int ?? 0) == vaultGeneration else { throw VaultError.staleGeneration }
        if let fingerprints = current["deep_link_fingerprints"] { record["deep_link_fingerprints"] = fingerprints }
        if http.statusCode == 401 || (terminalSession && (200..<300).contains(http.statusCode)) { try clearCredentials(&record) } else if issuedSession { try saveRecord(record) }
        let sanitized = String(data: try JSONSerialization.data(withJSONObject: envelope), encoding: .utf8) ?? ""
        return ["status": http.statusCode, "headers": ["Content-Type": http.value(forHTTPHeaderField: "Content-Type") ?? "application/json"], "body": sanitized, "generation": record["vault_generation"] as? Int ?? 0, "sessionChanged": http.statusCode == 401 || issuedSession || (terminalSession && (200..<300).contains(http.statusCode))]
    }

    private func rejectCredentialResidue(_ value: Any, depth: Int, visited: Int) throws -> Int {
        guard depth <= 16, visited < 10000 else { throw VaultError.invalidEnvelope }
        var count = visited + 1
        if let object = value as? [String: Any] {
            for (name, child) in object {
                guard !["token", "csrf_token", "guest_browser_nonce"].contains(name) else { throw VaultError.invalidEnvelope }
                count = try rejectCredentialResidue(child, depth: depth + 1, visited: count)
            }
        } else if let array = value as? [Any] {
            for child in array { count = try rejectCredentialResidue(child, depth: depth + 1, visited: count) }
        }
        return count
    }

    private func exactBackend(_ value: String) throws -> String {
        guard value == "https://casino.tiltseven.com", let components = URLComponents(string: value), components.scheme == "https", components.host == "casino.tiltseven.com", components.port == nil, components.user == nil, components.password == nil, components.path.isEmpty, components.query == nil, components.fragment == nil else { throw VaultError.invalidOrigin }
        return components.url!.absoluteString
    }

    private func exactWebView(_ value: String) throws -> String {
        guard let components = URLComponents(string: value), ["https", "capacitor"].contains(components.scheme), components.host == "localhost", components.port == nil, components.user == nil, components.password == nil, components.path.isEmpty, components.query == nil, components.fragment == nil else { throw VaultError.invalidOrigin }
        return components.url!.absoluteString
    }

    private func clearCredentials(_ record: inout [String: Any]) throws {
        ["token", "csrf_token", "guest_browser_nonce", "account_id", "session_id", "session_generation"].forEach { record.removeValue(forKey: $0) }
        record["vault_generation"] = (record["vault_generation"] as? Int ?? 0) + 1
        try saveRecord(record)
    }

    private func loadRecord() throws -> [String: Any] {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account, kSecReturnData as String: true, kSecMatchLimit as String: kSecMatchLimitOne]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecItemNotFound { return [:] }
        guard status == errSecSuccess, let data = item as? Data, let record = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw VaultError.invalidRecord }
        return record
    }

    private func saveRecord(_ record: [String: Any]) throws {
        let data = try JSONSerialization.data(withJSONObject: record)
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account]
        let attributes: [String: Any] = [kSecValueData as String: data, kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly, kSecAttrSynchronizable as String: false]
        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if status == errSecItemNotFound {
            var inserted = query
            attributes.forEach { inserted[$0.key] = $0.value }
            guard SecItemAdd(inserted as CFDictionary, nil) == errSecSuccess else { throw VaultError.keychainFailure }
        } else if status != errSecSuccess { throw VaultError.keychainFailure }
    }

    private enum VaultError: Error {
        case invalidRecord, invalidEnvelope, staleGeneration, invalidPath, invalidMethod, credentialHeader, bodyTooLarge, invalidOrigin, keychainFailure
    }
}

File: mobile/android/app/src/main/java/io/github/andreivorobiev/virtualcasino/CasinoSecureTransportPlugin.java
package io.github.andreivorobiev.virtualcasino;

import android.content.Context;
import android.content.SharedPreferences;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;
import android.util.Base64;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.KeyStore;
import java.util.Iterator;
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import org.json.JSONArray;
import org.json.JSONObject;

/** Owns native HTTPS and an Android-Keystore encrypted session record for issue #183. */
@CapacitorPlugin(name = "CasinoSecureTransport")
public final class CasinoSecureTransportPlugin extends Plugin {
    private static final String KEY_ALIAS = "casino.mobile.session.v1";
    private static final String PREFS_NAME = "casino_mobile_vault";
    private static final String RECORD_KEY = "encrypted_record";
    private static final int MAX_BODY_BYTES = 1_048_576;
    private static final java.util.Set<String> PUBLIC_HEADERS = new java.util.HashSet<>(java.util.Arrays.asList("accept", "content-type", "idempotency-key"));

    @PluginMethod
    public synchronized void configure(PluginCall call) {
        try {
            String backend = exactBackend(call.getString("backendOrigin", ""));
            String webView = exactWebView(call.getString("webViewOrigin", ""));
            JSONObject record = loadRecord();
            if (record.has("backend_origin") || record.has("webview_origin")) {
                if (!backend.equals(record.optString("backend_origin")) || !webView.equals(record.optString("webview_origin"))) call.reject("MOBILE_NATIVE_ORIGIN_REBIND_REJECTED"); else call.resolve();
                return;
            }
            record.put("backend_origin", backend);
            record.put("webview_origin", webView);
            record.put("vault_generation", record.optInt("vault_generation", record.optInt("generation", 0)));
            record.remove("generation");
            saveRecord(record);
            call.resolve();
        } catch (Exception error) {
            call.reject("MOBILE_NATIVE_CONFIGURATION_INVALID");
        }
    }

    @PluginMethod
    public void request(PluginCall call) {
        bridge.executeOnThread(() -> {
            try {
                Object requestedGeneration = call.getData().opt("generation");
                if (!(requestedGeneration instanceof Number) || ((Number) requestedGeneration).intValue() < 0 || ((Number) requestedGeneration).doubleValue() != ((Number) requestedGeneration).intValue()) throw new IllegalArgumentException("generation required");
                JSONObject result = perform(call.getString("path", ""), call.getString("method", "GET"), call.getObject("headers", new JSObject()), call.getString("body", ""), ((Number) requestedGeneration).intValue());
                call.resolve(JSObject.fromJSONObject(result));
            } catch (Exception error) {
                call.reject("MOBILE_NATIVE_REQUEST_FAILED");
            }
        });
    }

    @PluginMethod
    public void probe(PluginCall call) {
        bridge.executeOnThread(() -> {
            synchronized (this) {
                try {
                JSONObject record = loadRecord();
                if (record.optString("token").isEmpty()) {
                    JSObject absent = new JSObject();
                    absent.put("authenticated", false);
                    absent.put("generation", record.optInt("vault_generation", 0));
                    absent.put("status", 401);
                    call.resolve(absent);
                    return;
                }
                JSONObject response = perform("/api/v2/auth/mobile/session", "GET", new JSObject(), "", record.optInt("vault_generation", 0));
                int status = response.getInt("status");
                if (status != 200 && status != 401) throw new IllegalStateException("probe rejected");
                if (status == 200) {
                    JSONObject probeEnvelope = new JSONObject(response.getString("body"));
                    JSONObject probeData = probeEnvelope.optJSONObject("data");
                    JSONObject probeSession = probeData == null ? null : probeData.optJSONObject("session");
                    Object probeGeneration = probeSession == null ? null : probeSession.opt("generation");
                    int storedSessionGeneration = record.optInt("session_generation", -1);
                    if (!Boolean.TRUE.equals(probeData == null ? null : probeData.opt("authenticated")) || !(probeGeneration instanceof Number) || storedSessionGeneration < 1 || ((Number) probeGeneration).intValue() != storedSessionGeneration || ((Number) probeGeneration).doubleValue() != ((Number) probeGeneration).intValue() || !(probeSession.opt("issued_at") instanceof String) || ((String) probeSession.opt("issued_at")).isEmpty() || !(probeSession.opt("expires_at") instanceof String) || ((String) probeSession.opt("expires_at")).isEmpty() || !"active".equals(probeSession.opt("status"))) throw new IllegalStateException("probe envelope invalid");
                }
                JSObject result = new JSObject();
                result.put("authenticated", status == 200);
                result.put("generation", response.getInt("generation"));
                result.put("status", status);
                call.resolve(result);
                } catch (Exception error) {
                    call.reject("MOBILE_NATIVE_SESSION_PROBE_FAILED");
                }
            }
        });
    }

    @PluginMethod
    public void revokeAndClear(PluginCall call) {
        bridge.executeOnThread(() -> {
            try {
                synchronized (this) {
                    JSONObject record = loadRecord();
                    if (record.optString("token").isEmpty()) {
                        JSObject empty = new JSObject();
                        empty.put("revoked", true);
                        empty.put("cleared", true);
                        call.resolve(empty);
                        return;
                    }
                    JSONObject response = perform("/api/v2/auth/mobile/session/revoke", "POST", new JSObject(), "{}", record.optInt("vault_generation", 0));
                    if (response.getInt("status") != 200 && response.getInt("status") != 401) throw new IllegalStateException("revoke rejected");
                    JSObject result = new JSObject();
                    result.put("revoked", true);
                    result.put("cleared", true);
                    call.resolve(result);
                }
            } catch (Exception error) {
                call.reject("MOBILE_NATIVE_ACCOUNT_SWITCH_REVOKE_FAILED");
            }
        });
    }

    @PluginMethod
    public synchronized void claimDeepLink(PluginCall call) {
        try {
            String fingerprint = call.getString("fingerprint", "");
            if (!fingerprint.matches("^[a-f0-9]{64}$")) throw new IllegalArgumentException("invalid fingerprint");
            JSONObject record = loadRecord();
            JSONArray retained = record.optJSONArray("deep_link_fingerprints");
            if (retained == null) retained = new JSONArray();
            for (int index = 0; index < retained.length(); index += 1) {
                if (fingerprint.equals(retained.optString(index))) {
                    JSObject replay = new JSObject();
                    replay.put("claimed", false);
                    call.resolve(replay);
                    return;
                }
            }
            JSONArray bounded = new JSONArray();
            for (int index = Math.max(0, retained.length() - 63); index < retained.length(); index += 1) bounded.put(retained.getString(index));
            bounded.put(fingerprint);
            record.put("deep_link_fingerprints", bounded);
            saveRecord(record);
            JSObject result = new JSObject();
            result.put("claimed", true);
            call.resolve(result);
        } catch (Exception error) {
            call.reject("MOBILE_NATIVE_DEEP_LINK_CLAIM_FAILED");
        }
    }

    private synchronized JSONObject perform(String path, String method, JSObject publicHeaders, String body, int expectedGeneration) throws Exception {
        JSONObject record = loadRecord();
        int vaultGeneration = record.optInt("vault_generation", 0);
        if (expectedGeneration < 0 || expectedGeneration != vaultGeneration) throw new IllegalStateException("stale generation");
        if (!path.matches("^/api/(?:v1|v2)/[A-Za-z0-9_./?=&%+-]*$") || path.contains("//")) throw new IllegalArgumentException("invalid path");
        String normalizedMethod = method.toUpperCase();
        if (!normalizedMethod.matches("GET|POST|PUT|PATCH|DELETE|HEAD")) throw new IllegalArgumentException("invalid method");
        if (("/api/v2/auth/login".equals(path) || "/api/v2/auth/guest".equals(path)) && !record.optString("token").isEmpty()) throw new IllegalStateException("active session replacement rejected");
        URL url = new URL(record.getString("backend_origin") + path);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        try {
        connection.setInstanceFollowRedirects(false);
        connection.setUseCaches(false);
        connection.setConnectTimeout(10000);
        connection.setReadTimeout(15000);
        connection.setRequestMethod(normalizedMethod);
        connection.setRequestProperty("Origin", record.getString("webview_origin"));
        connection.setRequestProperty("X-Casino-Mobile-Client", "1");
        Iterator<String> headerNames = publicHeaders.keys();
        while (headerNames.hasNext()) {
            String name = headerNames.next();
            String lower = name.toLowerCase();
            if (!PUBLIC_HEADERS.contains(lower)) throw new IllegalArgumentException("unowned header");
            connection.setRequestProperty(name, publicHeaders.getString(name));
        }
        connection.setRequestProperty("Cache-Control", "no-store");
        String token = record.optString("token");
        String csrf = record.optString("csrf_token");
        if (!token.isEmpty()) connection.setRequestProperty("Authorization", "Bearer " + token);
        if (!csrf.isEmpty() && normalizedMethod.matches("POST|PUT|PATCH|DELETE")) connection.setRequestProperty("X-CSRF-Token", csrf);
        String guestNonce = record.optString("guest_browser_nonce");
        if (!guestNonce.isEmpty()) connection.setRequestProperty("X-Guest-Browser-Nonce", guestNonce);
        if ("/api/v2/auth/mobile/session/rotate".equals(path)) body = new JSONObject().put("expected_generation", record.optInt("session_generation", -1)).toString();
        if (!body.isEmpty() && normalizedMethod.matches("POST|PUT|PATCH|DELETE")) {
            byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
            if (bytes.length > MAX_BODY_BYTES) throw new IllegalArgumentException("body too large");
            connection.setDoOutput(true);
            try (OutputStream output = connection.getOutputStream()) { output.write(bytes); }
        }
        int status = connection.getResponseCode();
        InputStream stream = status >= 400 ? connection.getErrorStream() : connection.getInputStream();
        String responseBody = readBounded(stream);
        if (responseBody.isEmpty()) throw new IllegalStateException("empty API response");
        JSONObject envelope = new JSONObject(responseBody);
        JSONObject data = envelope.optJSONObject("data");
        boolean successStatus = status >= 200 && status < 300;
        if (successStatus && (!Boolean.TRUE.equals(envelope.opt("ok")) || data == null)) throw new IllegalStateException("success envelope invalid");
        boolean terminalSession = "/api/v2/auth/logout".equals(path) || "/api/v2/auth/guest/end".equals(path) || "/api/v2/auth/mobile/session/revoke".equals(path);
        if (terminalSession && successStatus) {
            String acknowledgement = "/api/v2/auth/logout".equals(path) ? "logged_out" : ("/api/v2/auth/guest/end".equals(path) ? "ended" : "revoked");
            if (!Boolean.TRUE.equals(data.opt(acknowledgement))) throw new IllegalStateException("terminal acknowledgement invalid");
        }
        boolean credentialIssuance = ("/api/v2/auth/login".equals(path) || "/api/v2/auth/guest".equals(path) || "/api/v2/auth/mobile/session/rotate".equals(path)) && successStatus;
        JSONObject session = data == null ? null : data.optJSONObject("session");
        boolean issuedSession = false;
        if (credentialIssuance && session != null) {
            Object issuedToken = session.remove("token");
            Object issuedCsrf = session.remove("csrf_token");
            Object issuedSessionId = session.opt("session_id");
            Object issuedGeneration = session.opt("generation");
            if (!(issuedToken instanceof String) || ((String) issuedToken).isEmpty() || !(issuedCsrf instanceof String) || ((String) issuedCsrf).isEmpty() || !(issuedSessionId instanceof String) || ((String) issuedSessionId).isEmpty() || !(issuedGeneration instanceof Number) || ((Number) issuedGeneration).intValue() < 1 || ((Number) issuedGeneration).doubleValue() != ((Number) issuedGeneration).intValue()) throw new IllegalStateException("issuance invalid");
            record.put("token", issuedToken);
            record.put("csrf_token", issuedCsrf);
            record.put("session_id", issuedSessionId);
            record.put("session_generation", ((Number) issuedGeneration).intValue());
            record.put("vault_generation", vaultGeneration + 1);
            JSONObject user = data.optJSONObject("user");
            String accountId = user == null ? data.optString("user_id") : user.optString("user_id");
            if (!accountId.isEmpty()) record.put("account_id", accountId);
            issuedSession = true;
        }
        if (credentialIssuance && !issuedSession) throw new IllegalStateException("issuance invalid");
        String issuedGuestNonce = "";
        if (credentialIssuance && "/api/v2/auth/guest".equals(path) && data != null) {
            Object guestNonceValue = data.remove("guest_browser_nonce");
            issuedGuestNonce = guestNonceValue instanceof String ? (String) guestNonceValue : "";
            if (!issuedGuestNonce.isEmpty()) record.put("guest_browser_nonce", issuedGuestNonce);
        }
        if ("/api/v2/auth/guest".equals(path) && credentialIssuance && issuedGuestNonce.isEmpty()) throw new IllegalStateException("guest issuance invalid");
        rejectCredentialResidue(envelope, 0, new int[] {0});
        if (status == 401 || (terminalSession && status >= 200 && status < 300)) clearCredentials(record); else if (issuedSession) saveRecord(record);
        JSObject headers = new JSObject();
        String contentType = connection.getHeaderField("Content-Type");
        if (contentType != null) headers.put("Content-Type", contentType);
        JSONObject result = new JSONObject();
        result.put("status", status);
        result.put("headers", headers);
        result.put("body", envelope.length() == 0 ? responseBody : envelope.toString());
        result.put("generation", loadRecord().optInt("vault_generation", 0));
        result.put("sessionChanged", status == 401 || issuedSession || (terminalSession && status >= 200 && status < 300));
        return result;
        } finally {
            connection.disconnect();
        }
    }

    private void rejectCredentialResidue(Object value, int depth, int[] visited) throws Exception {
        if (depth > 16 || ++visited[0] > 10000) throw new IllegalStateException("response shape invalid");
        if (value instanceof JSONObject) {
            JSONObject object = (JSONObject) value;
            Iterator<String> names = object.keys();
            while (names.hasNext()) {
                String name = names.next();
                if ("token".equals(name) || "csrf_token".equals(name) || "guest_browser_nonce".equals(name)) throw new IllegalStateException("credential residue");
                rejectCredentialResidue(object.get(name), depth + 1, visited);
            }
        } else if (value instanceof JSONArray) {
            JSONArray array = (JSONArray) value;
            for (int index = 0; index < array.length(); index += 1) rejectCredentialResidue(array.get(index), depth + 1, visited);
        }
    }

    private String readBounded(InputStream stream) throws Exception {
        if (stream == null) return "";
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[8192];
        int total = 0;
        int count;
        while ((count = stream.read(buffer)) != -1) {
            total += count;
            if (total > MAX_BODY_BYTES) throw new IllegalStateException("response too large");
            output.write(buffer, 0, count);
        }
        return output.toString(StandardCharsets.UTF_8.name());
    }

    private String exactBackend(String value) throws Exception {
        URI uri = new URI(value);
        if (!"https://casino.tiltseven.com".equals(value) || !"https".equals(uri.getScheme()) || !"casino.tiltseven.com".equals(uri.getHost()) || uri.getUserInfo() != null || uri.getPort() != -1 || uri.getPath() == null || !uri.getPath().isEmpty() || uri.getQuery() != null || uri.getFragment() != null) throw new IllegalArgumentException("backend");
        return uri.toString();
    }

    private String exactWebView(String value) throws Exception {
        URI uri = new URI(value);
        if (!("https".equals(uri.getScheme()) || "capacitor".equals(uri.getScheme())) || !"localhost".equals(uri.getHost()) || uri.getUserInfo() != null || uri.getPort() != -1 || (uri.getPath() != null && !uri.getPath().isEmpty()) || uri.getQuery() != null || uri.getFragment() != null) throw new IllegalArgumentException("webview");
        return uri.toString();
    }

    private synchronized void clearCredentials(JSONObject record) throws Exception {
        record.remove("token");
        record.remove("csrf_token");
        record.remove("guest_browser_nonce");
        record.remove("account_id");
        record.remove("session_id");
        record.remove("session_generation");
        record.put("vault_generation", record.optInt("vault_generation", 0) + 1);
        saveRecord(record);
    }

    private SharedPreferences preferences() {
        return getContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    private SecretKey key() throws Exception {
        KeyStore store = KeyStore.getInstance("AndroidKeyStore");
        store.load(null);
        if (store.containsAlias(KEY_ALIAS)) return ((KeyStore.SecretKeyEntry) store.getEntry(KEY_ALIAS, null)).getSecretKey();
        KeyGenerator generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
        generator.init(new KeyGenParameterSpec.Builder(KEY_ALIAS, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT).setBlockModes(KeyProperties.BLOCK_MODE_GCM).setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE).setKeySize(256).build());
        return generator.generateKey();
    }

    private synchronized JSONObject loadRecord() throws Exception {
        String encoded = preferences().getString(RECORD_KEY, "");
        if (encoded.isEmpty()) return new JSONObject();
        byte[] packed = Base64.decode(encoded, Base64.NO_WRAP);
        if (packed.length < 13) throw new IllegalStateException("vault record invalid");
        byte[] iv = new byte[12];
        byte[] ciphertext = new byte[packed.length - 12];
        System.arraycopy(packed, 0, iv, 0, 12);
        System.arraycopy(packed, 12, ciphertext, 0, ciphertext.length);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, key(), new GCMParameterSpec(128, iv));
        return new JSONObject(new String(cipher.doFinal(ciphertext), StandardCharsets.UTF_8));
    }

    private synchronized void saveRecord(JSONObject record) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, key());
        byte[] ciphertext = cipher.doFinal(record.toString().getBytes(StandardCharsets.UTF_8));
        byte[] iv = cipher.getIV();
        byte[] packed = new byte[iv.length + ciphertext.length];
        System.arraycopy(iv, 0, packed, 0, iv.length);
        System.arraycopy(ciphertext, 0, packed, iv.length, ciphertext.length);
        if (!preferences().edit().putString(RECORD_KEY, Base64.encodeToString(packed, Base64.NO_WRAP)).commit()) throw new IllegalStateException("vault commit failed");
    }
}
